### PR TITLE
Add the function to generate metasploit-framework module list

### LINKED ISSUE #2: Notify a status of GitHub Actions job

### PR DESCRIPTION
**Pull request by takuzoo3868**
This PR will be used for CVEs in the metasploit-framework-db detection of **Vuls** in the future. 
The functions provided are as follows.

# What did you implement:

- Fetch the latest metasploit-framework to the cache directory
- Update msfdb-list which stores CVEs list
- Extracting the necessary information from the metasploit-framework module scripts

Modules in the CVE are biased towards **auxiliary & exploits**, 
so I added the updater code to these two only.

**NOTICE**: Push to msfdb-list repo will be implemented after the review...

# How Has This Been Tested?

The samples are used to test the regular expression-related parts of the code.

```
$ make test
fatal: No names found, cannot describe anything.
GO111MODULE=off go get -u golang.org/x/lint/golint
golint github.com/vulsio/msfdb-list-updater github.com/vulsio/msfdb-list-updater/git github.com/vulsio/msfdb-list-updater/log github.com/vulsio/msfdb-list-updater/models github.com/vulsio/msfdb-list-updater/msf github.com/vulsio/msfdb-list-updater/utils
echo github.com/vulsio/msfdb-list-updater github.com/vulsio/msfdb-list-updater/git github.com/vulsio/msfdb-list-updater/log github.com/vulsio/msfdb-list-updater/models github.com/vulsio/msfdb-list-updater/msf github.com/vulsio/msfdb-list-updater/utils | xargs env GO111MODULE=on go vet || exit;
gofmt -d git/git.go; gofmt -d log/log.go; gofmt -d main.go; gofmt -d models/type.go; gofmt -d msf/msf.go; gofmt -d msf/msf_test.go; gofmt -d utils/last_updated.go; gofmt -d utils/utils.go;
diff -u models/type.go.orig models/type.go
--- models/type.go.orig	2020-06-14 15:05:21.000000000 +0900
+++ models/type.go	2020-06-14 15:05:21.000000000 +0900
@@ -11,4 +11,4 @@
 }

 // Modules :
-type Modules []Module
\ No newline at end of file
+type Modules []Module
GO111MODULE=on go test -cover -v ./... || exit;
?   	github.com/vulsio/msfdb-list-updater	[no test files]
?   	github.com/vulsio/msfdb-list-updater/git	[no test files]
?   	github.com/vulsio/msfdb-list-updater/log	[no test files]
?   	github.com/vulsio/msfdb-list-updater/models	[no test files]
=== RUN   TestParse
=== RUN   TestParse/example.rb
=== RUN   TestParse/example_webapp.rb
=== RUN   TestParse/binder_uaf.rb
=== RUN   TestParse/example.rb#01
=== RUN   TestParse/openemr_sqli_dump.rb
--- PASS: TestParse (0.01s)
    --- PASS: TestParse/example.rb (0.00s)
    --- PASS: TestParse/example_webapp.rb (0.00s)
    --- PASS: TestParse/binder_uaf.rb (0.00s)
    --- PASS: TestParse/example.rb#01 (0.00s)
    --- PASS: TestParse/openemr_sqli_dump.rb (0.00s)
PASS
coverage: 47.6% of statements
ok  	github.com/vulsio/msfdb-list-updater/msf	(cached)	coverage: 47.6% of statements
?   	github.com/vulsio/msfdb-list-updater/utils	[no test files]
```

# Others

### Usage

```
$ msfdb-list-updater -h
Usage of msfdb-list-updater:
  -target string
        update target (msf)
```

スクリーンショット 2020-06-14 15 11 32

### msfdb-list
By executing updater, the msfdb-list is changed as follows.

```
$ tree -L 2 msfdb-list/
msfdb-list/
├── README.md
├── last_updated.json
└── rapid7
    ├── 1999
    ├── 2000
    ├── 2001
    ├── 2002
    ├── 2003
    ├── 2004
    ├── 2005
    ├── 2006
    ├── 2007
    ├── 2008
    ├── 2009
    ├── 2010
    ├── 2011
    ├── 2012
    ├── 2013
    ├── 2014
    ├── 2015
    ├── 2016
    ├── 2017
    ├── 2018
    ├── 2019
    └── 2020

23 directories, 2 files
```

```
$ bat msfdb-list/rapid7/2020/CVE-2020-4429.json
───────┬──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
       │ File: msfdb-list/rapid7/2020/CVE-2020-4429.json
───────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1   │ [
   2   │   {
   3   │     "Name": "ibm_drm_download.rb",
   4   │     "Title": "IBM Data Risk Manager Arbitrary File Download",
   5   │     "Discription": "IBM Data Risk Manager (IDRM) contains two vulnerabilities that can be chained by an unauthenticated attacker to download arbitrary files off the system. The first is an unauthenticated bypass, followed by a pa
       │ th traversal. This module exploits both vulnerabilities, giving an attacker the ability to download (non-root) files. A downloaded file is zipped, and this module also unzips it before storing it in the database. By default this
       │ module downloads Tomcat's application.properties files, which contains the database password, amongst other sensitive data. At the time of disclosure, this is a 0 day. Versions 2.0.3 and 2.0.2 are confirmed to be affected, and th
       │ e latest 2.0.6 is most likely affected too. Version 2.0.1 is not vulnerable.",
   6   │     "CveIDs": [
   7   │       "CVE-2020-4427",
   8   │       "CVE-2020-4429"
   9   │     ],
  10   │     "References": [
  11   │       "https://github.com/pedrib/PoC/blob/master/advisories/IBM/ibm_drm/ibm_drm_rce.md",
  12   │       "https://seclists.org/fulldisclosure/2020/Apr/33"
  13   │     ]
  14   │   },
  15   │   {
  16   │     "Name": "ibm_drm_rce.rb",
  17   │     "Title": "IBM Data Risk Manager Unauthenticated Remote Code Execution",
  18   │     "Discription": "IBM Data Risk Manager (IDRM) contains three vulnerabilities that can be chained by an unauthenticated attacker to achieve remote code execution as root. The first is an unauthenticated bypass, followed by a co
       │ mmand injection as the server user, and finally abuse of an insecure default password. This module exploits all three vulnerabilities, giving the attacker a root shell. At the time of disclosure, this is a 0day. Versions 2.0.3 an
       │ d below are confirmed to be affected, and the latest 2.0.6 is most likely affected too.",
  19   │     "CveIDs": [
  20   │       "CVE-2020-4427",
  21   │       "CVE-2020-4428",
  22   │       "CVE-2020-4429"
  23   │     ],
  24   │     "References": [
  25   │       "https://github.com/pedrib/PoC/blob/master/advisories/IBM/ibm_drm/ibm_drm_rce.md",
  26   │       "https://seclists.org/fulldisclosure/2020/Apr/33"
  27   │     ]
  28   │   },
  29   │   {
  30   │     "Name": "ibm_drm_a3user.rb",
  31   │     "Title": "IBM Data Risk Manager a3user Default Password",
  32   │     "Discription": "This module abuses a known default password in IBM Data Risk Manager. The 'a3user' has the default password 'idrm' and allows an attacker to log in to the virtual appliance via SSH. This can be escalate to ful
       │ l root access, as 'a3user' has sudo access with the default password. At the time of disclosure, this is a 0day. Versions <= 2.0.3 are confirmed to be affected, and the latest 2.0.6 is most likely affected too.",
  33   │     "CveIDs": [
  34   │       "CVE-2020-4429"
  35   │     ],
  36   │     "References": [
  37   │       "https://github.com/pedrib/PoC/blob/master/advisories/IBM/ibm_drm/ibm_drm_rce.md",
  38   │       "https://seclists.org/fulldisclosure/2020/Apr/33"
  39   │     ]
  40   │   }
  41   │ ]
───────┴──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```

### CODE REVIEW

File: utils/utils.go [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+// LookupEnv :
+func LookupEnv(key, defaultValue string) string {
+	if val, ok := os.LookupEnv(key); ok {
+		return val
+	}
+	return defaultValue
+}
+
+// SaveCVEPerYear :
+func SaveCVEPerYear(dirName string, cveID string, data interface{}) error {
```
Comment (kotakanbe) [inline utils/utils.go]: Why is data an interface, and why not a struct? 
I think a struct doesn't require any casting and is more maintainable.

File: utils/utils.go [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+		return err
+	}
+	defer f.Close()
+
+	b, err := json.MarshalIndent(data, "", "  ")
+	if err != nil {
+		return err
+	}
+
+	b = bytes.Replace(b, []byte("\\u003c"), []byte("<"), -1)
```
Comment (kotakanbe) [inline utils/utils.go]: We should use API
https://stackoverflow.com/a/39079218

File: msf/msf.go [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+		Name:        filepath.Base(path),
+		Title:       title,
+		Description: decs,
+		CveIDs:      cveIDs,
+		EdbIDs:      edbIDs,
+		References:  links,
+	}, nil
+}
+
+func formatDescription(match [][]byte) string {
```
Comment (kotakanbe) [inline msf/msf.go]: Please add a test case for this function.

**Review by kotakanbe — CHANGES_REQUESTED** [review]

**Comment by kotakanbe** [comment]
Please check out some of the comments I made :)

File: msf/msf.go (hunk 3 above)
Comment (takuzoo3868) [reply msf/msf.go]: This function is included in the Parse test.   
Should I set up a separate test case to check it?

**Review by kotakanbe — APPROVED** [review]
LGTM

**Pull request merged**